sling/nlp/parser/tools/train_pytorch.py

```diff
@@ -0,0 +1,377 @@
+    def set(self, args):
+      if args.learning_rate is not None:
+        self.lr = args.learning_rate
+      if args.batch_size is not None:
+        self.batch_size = args.batch_size
+      if args.train_steps is not None:
+        self.num_examples = args.train_steps * self.batch_size
+      if args.report_every is not None:
+        self.report_every = args.report_every * self.batch_size
+      if args.l2_coeff is not None:
+        self.l2_coeff = args.l2_coeff
+      if args.gradient_clip_norm is not None:
+        self.gradient_clip = args.gradient_clip_norm
+      if args.learning_method is not None:
+        self.optimizer = args.learning_method
+      if args.adam_beta1 is not None:
+        self.adam_beta1 = args.adam_beta1
+      if args.adam_beta2 is not None:
+        self.adam_beta2 = args.adam_beta2
+      if args.adam_eps is not None:
+        self.adam_eps = args.adam_eps
+      if args.use_moving_average is not None:
+        self.moving_avg = args.use_moving_average
+      if args.moving_average_coeff is not None:
+        self.moving_avg_coeff = args.moving_average_coeff
+
+
+    # Returns a string representation of all hyperparameters.
+    def __str__(self):
+      return str(self.__dict__)
+
+
+  # Instantiates the trainer with the given model, optional evaluator,
+  # and hyperparameters.
+  def __init__(self, sempar, evaluator=None, output_file_prefix=None, \
+               hyperparams=Hyperparams()):
+    self.model = sempar
+    self.evaluator = evaluator
+    self.hyperparams = hyperparams
+
+    if hyperparams.optimizer == "sgd":
+      self.optimizer = torch.optim.SGD(
+        sempar.parameters(),
+        lr=self.hyperparams.lr,
+        momentum=0,
+        dampening=0,
+        weight_decay=0,
+        nesterov=False)
+    elif hyperparams.optimizer == "adam":
+      self.optimizer = torch.optim.Adam(
+          sempar.parameters(), lr=hyperparams.lr, weight_decay=0, \
+              betas=(hyperparams.adam_beta1, hyperparams.adam_beta2), \
+              eps=hyperparams.adam_eps)
+
+    num_params = 0
+    for name, p in sempar.named_parameters():
+      if p.requires_grad:
+        print name, ":", p.size()
+        num_params += torch.numel(p)
+    print "Number of parameters:", num_params
+
+    self.count = 0
+    self.last_eval_count = 0
+
+    self.current_batch_num_transitions = 0
+    self.current_batch_size = 0
+    self.batch_loss = Var(torch.FloatTensor([0.0]))
+    self._reset()
+
+    self.checkpoint_metrics = []
+    self.best_metric = None
+    self.output_file_prefix = output_file_prefix
+
+    # Exponential moving average clones.
+    self.averages = {}
+    if hyperparams.moving_avg:
+      for name, p in sempar.named_parameters():
+        if p.requires_grad:
+          self.averages[name] = p.data.clone()
+
+
+  # Resets the state for a new batch.
+  def _reset(self):
+    self.current_batch_size = 0
+    self.current_batch_num_transitions = 0
+    del self.batch_loss
```

Comment:
Why do you need a del here? You are overriding this on the next line.

Reply:
self.batch_loss is the apex of the computation graph. I noticed that if I just reset it (as I do on the next line) and not call del, Python was too lazy to garbage collect the old batch loss (and thus the entire graph). I know that with del, Python might still not return the freed memory to the OS (at least immediately), but empirically I recall seeing much better memory behavior.

Reply:
It might be good to document this in the code since this is not obvious. AFAIK there should normally not be any difference between del and overwriting